File: 2017/cpp/day02.cpp
#include <iostream>
#include <sstream>
#include <limits>
#include <vector>
#include <iterator>
using namespace std;

int part1(vector<string> lines) {
  int total = 0;

  for (auto line : lines) {
    istringstream is(line);
    int max = numeric_limits<int>::min();
    int min = numeric_limits<int>::max();
    for (int i; is >> i;) {
      if (i > max) {
        max = i;
      }
      if (i < min) {
        min = i;
      }
    }
    total += max - min;
  }

  return total;
}

int part2(vector<string> lines) {
  vector<vector<int>> data;

  for (auto line : lines) {
    istringstream is(line);
    data.push_back(vector<int>((istream_iterator<int>(is)), istream_iterator<int>()));
  }

  int total = 0;
  for (auto v : data) {
    for (auto a : v) {
      for (auto b : v) {
        if (a != b && a % b == 0) {
          total += a / b;
        }
      }
    }
  }

  return total;
}

int main() {
  vector<string> lines;
  for (string line; getline(cin, line);) {
    lines.push_back(line);
  }

  cout << "Part 1: " << part1(lines) << "\n";
  cout << "Part 2: " << part2(lines) << "\n";
}

File: 2017/rust/src/bin/day02.rs
use std::io;
use std::io::prelude::*;

fn main() {
    let parse_line = |l: String| {
        l.split_whitespace().filter_map(|s| s.parse().ok()).collect()
    };

    let lines: Vec<Vec<i32>> = io::stdin().lock()
        .lines()
        .filter_map(|l| l.map(parse_line).ok())
        .collect();

    println!("Part 1: {}", part1(&lines));
    println!("Part 2: {}", part2(&lines));
}

fn part1(lines: &Vec<Vec<i32>>) -> i32 {
    let mut total = 0;
    for line in lines {
        let mut min = std::i32::MAX;
        let mut max = std::i32::MIN;
        for &n in line {
            if n < min { min = n }
            if n > max { max = n }
        }
        total += max - min;
    }
    total
}

fn part2(lines: &Vec<Vec<i32>>) -> i32 {
    let mut total = 0;

    for line in lines {
        for &a in line {
            for &b in line {
                if a != b && a % b == 0 {
                    total += a / b;
                }
            }
        }
    }

    total
}
